sim/paladin/divine_favor.go

```diff
@@ -0,0 +1,69 @@
+package paladin
+
+import (
+	"time"
+
+	"github.com/wowsims/sod/sim/core"
+)
+
+func (paladin *Paladin) registerDivineFavorSpellAndAura() {
+	if !paladin.Talents.DivineFavor {
+		return
+	}
+	var affectedSpells []*core.Spell
+
+	cdTimer := paladin.NewTimer()
+	cd := time.Minute * 2
+
+	auraActionID := core.ActionID{SpellID: 20216}
+
+	paladin.DivineFavorAura = paladin.RegisterAura(core.Aura{
+		Label:    "Divine Favor",
+		ActionID: auraActionID,
+		Duration: core.NeverExpires,
+		OnInit: func(aura *core.Aura, sim *core.Simulation) {
+			affectedSpells = core.FilterSlice(
+				core.Flatten([][]*core.Spell{
+					paladin.HolyShock,
+				}), func(spell *core.Spell) bool { return spell != nil },
+			)
+		},
+		OnGain: func(aura *core.Aura, sim *core.Simulation) {
+			core.Each(affectedSpells, func(spell *core.Spell) {
+				spell.BonusCritRating += core.CritRatingPerCritChance * 100
+			})
+		},
+		OnExpire: func(aura *core.Aura, sim *core.Simulation) {
+			core.Each(affectedSpells, func(spell *core.Spell) {
+				spell.BonusCritRating -= core.CritRatingPerCritChance * 100
+			})
+		},
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if spell.SpellCode != SpellCode_PaladinHolyShock {
+				return
+			}
+			// Remove the buff and put skill on CD
+			aura.Deactivate(sim)
+			cdTimer.Set(sim.CurrentTime + cd)
```

Comment:
you should not need to set this timer manually if you're already passing it as the CD timer of the spell

Reply:
The Divine Favor spell only goes on cooldown once the Divine Favor aura expires, i.e. the player spends it on a Holy Shock cast.

The point of this code is to only set the cooldown running on the Divine Favor spell when the Holy Shock is cast. If I remove this line of code, Divine Favor gets cast every 2 mins on cooldown regardless of when the aura actually expired.

Is there a more elegant way to attain this behaviour that I have missed?

Reply:
No in this case what you did makes sense